doc/source/ray-core/internals/autoscaler-v2.rst

```diff
@@ -0,0 +1,204 @@
+
+The entry point of the autoscaler is `monitor.py <https://github.com/ray-project/ray/blob/03491225d59a1ffde99c3628969ccf456be13efd/python/ray/autoscaler/v2/monitor.py#L332>`__, which starts a GCS client and runs the reconciliation loop.
+
+This process is launched on the head node by the `start_head_processes <https://github.com/ray-project/ray/blob/03491225d59a1ffde99c3628969ccf456be13efd/python/ray/_private/node.py#L1439>`__ function when using the ``ray up`` cluster launcher.
+When running under KubeRay, it instead runs as a separate autoscaler container in the Head Pod.
+
+.. note::
+
+   In the case of the cluster launcher, if the autoscaler process crashes, then there is no autoscaling.
+   While in the case of KubeRay, Kubernetes restarts the autoscaler container if it crashes.
+
+
+The process periodically `reconciles <https://github.com/ray-project/ray/blob/03491225d59a1ffde99c3628969ccf456be13efd/python/ray/autoscaler/v2/autoscaler.py#L200-L213>`__ against a snapshot of the following information using the Reconciler:
+
+1. **The latest pending demands** (queried from the `get_cluster_resource_state` GCS RPC): Pending Ray tasks, actors, and placement groups.
+2. **The latest user cluster constraints** (queried from the `get_cluster_resource_state` GCS RPC): The minimum cluster size, if specified via the ``ray.autoscaler.sdk.request_resources`` invocation.
+3. **The latest total and available node resources** (queried from the `get_cluster_resource_state` GCS RPC): The total and currently available resources of each alive Ray node.
+4. **The latest cloud instances** (queried from the cloud provider's implementation): The list of instances managed by the cloud provider implementation.
+5. **The latest worker group configurations** (queried from the cluster YAML file or the RayCluster CRD).
+
+The preceding information is retrieved at the beginning of each reconciliation loop.
+The Reconciler uses this information to construct its internal state. This is the `sync phase <https://github.com/ray-project/ray/blob/03491225d59a1ffde99c3628969ccf456be13efd/python/ray/autoscaler/v2/instance_manager/reconciler.py#L112-L120>`__.
+
+After the sync phase, the Reconciler performs the `following steps <https://github.com/ray-project/ray/blob/03491225d59a1ffde99c3628969ccf456be13efd/python/ray/autoscaler/v2/scheduler.py#L840>`__ with the ``ResourceDemandScheduler``:
+
+1. Enforce configuration constraints, including min/max nodes for each worker group.
+2. Enforce user cluster constraints (if specified by `ray.autoscaler.sdk.request_resources` invocation).
+3. Fit pending demands into available resources on the cluster snapshot. This is the simulation mentioned earlier.
+4. Fit any remaining demands (left over from the previous step) against worker group configurations to determine which nodes to launch.
+5. Terminate idle instances according to each node's ``idle_duration_ms`` (queried from GCS) and the configured idle timeout for each group.
```

Comment:
If the idle node is needed by the previous steps then it won't be terminated right, worth mentioning here as well

Reply:
👌